design-documents/jwt-support.md

```diff
@@ -0,0 +1,180 @@
+
+And in general JWT looks like this: `header`.`payload`.`signature`.
+
+The JWT usage became more and more popular for authentication and data verification purposes and Magento has a lot of integrations with different 3rd party systens, we should support JWT creation/validation/parsing out-of-box. Another benefit, that JWT is language agnostic and token generaten with one programming language can be parsed with another, the only key should be shared between applications.
+
+## Solution
+
+We do not need to implement own solution for key generation, parsing tokens, encryption/decryption, etc. [There are](https://jwt.io/#libraries) multiple PHP libraries which support all needed operations and algorithms and we need to create own wrappers.
+
+[PHP JWT Framework](https://github.com/web-token/jwt-framework) - is proposed as a library because:
+ - implements all algorithms from RFC
+ - provides implementation for JWS, JWT, JWE, JWA, JWK, JSON Web Key Thumbprint, Unencoded Payload Option
+ - supports different serialization modes
+ - supports multiple compression methods
+ - full support of JSON Web Key Set
+ - has a good documentation
+ - supports detached payload, multiple signatures, nested tokens
+
+## Implementation
+
+The following diagram represents needed interfaces to unify the workflow for JWT usage.
+
+![Class diagram](img/jwt-class-diagram.png)
+
+The `\Magento\Framework\Jwt\KeyGeneratorInterface` will provide a possibility to use different types of key generators like: Magento deployment secret key, X.509 certificates or RSA keys.
+```php
+interface KeyGeneratorInterface
+{
+    public function create(): JWK;
+}
+
+class CryptKeyGenerator implements KeyGeneratorInterface
+{
+    public function __construct(AlgorithmFactory $algorithmFactory, DeploymentConfig $deploymentConfig)
+    {
+        $this->algorithmFactory = $algorithmFactory;
+        $this->deploymentConfig = $deploymentConfig;
+    }
+
+    public function create(): JWK
+    {
+        $secret = (string) $this->deploymentConfig->get('crypt/key');
+        return JWKFactory::createFromSecret(
+            $secret,
+            ['alg' => $this->algorithmFactory->getAlgorithmName(), 'use' => 'sig']
+        );
+    }
+}
+```
+
+The `\Magento\Framework\Jwt\GeneratorInterface` might be used to generate JWT based on claims:
+```php
+interface GeneratorInterface
+{
+    public function generate(array $claims = []): string;
+}
+
+class JwsGenerator implements GeneratorInterface
+{
+    public function __construct(
+        AlgorithmFactory $algorithmFactory,
+        CryptKeyGenerator $keyGenerator,
+        SerializerInterface $serializer
+    ) {
+        $this->algorithmFactory = $algorithmFactory;
+        $this->keyGenerator = $keyGenerator;
+        $this->serializer = $serializer;
+    }
+
+    public function generate(array $claims = []): string
+    {
+        $timestamp = time();
+        $baseClaims = [
```

Comment:
are these claims mandatory by the RFC? What if someone doesn't want them in their tokens?

Reply:
RFC doesn't describe mandatory claims, it describes only reserved claims. This is just an example from POC, the claims should be configurable.

Reply:
Updated